### PR TITLE
[SDK] Relax k8s sanitization

### LINKED ISSUE #2570: ParallelFor parameter parsing does not retain user input format
When running a pipeline with a ParallelFor condition over a list of dictionaries, camelCase and snake_case inputs are altered.

This won't work:

    with kfp.dsl.ParallelFor([{"helloWorld": "blah"}]) as item:
        my_op(item.helloWorld)

    with kfp.dsl.ParallelFor([{"hello_world": "blah"}]) as item:
        my_op(item.hello_world)

If I am reading this correctly, then the `my_op` is parsing `item.hello_world` as `item.hello-world` (and `helloWorld` as `helloworld`. However, the ParallelFor operation isn't  performing the same parsing. (I presume it is delivering `item.hello_world`, exactly as written in the dictionary, but I cannot confirm this).

The following works correctly:

    with kfp.dsl.ParallelFor([{"helloworld": "blah"}]) as item:
        my_op(item.helloWorld)

    with kfp.dsl.ParallelFor([{"hello-world": "blah"}]) as item:
        my_op(item.hello_world)

### PR DESCRIPTION
**Pull request by numerology**
Let k8s sanitization allows capital letters and under score in the names of parameters/outputs. This change will allow users to refer to self-defined attributes of a complex structure in a more Pythonic way. 

One implication of this change would be, we have to delegate to users the responsibility of ensuring the argo naming convention when they create any k8s object within the pipeline. For example, in the [volume ops example](https://github.com/kubeflow/pipelines/tree/master/samples/core/volume_ops), a user should specify `name='create-pvc'` instead of `name='create_pvc'` to comply with argo naming convention.

Fixes #2570

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2634)

### CODE REVIEW

**Comment by numerology** [comment]
/hold cancel
Seems to be okay for now.

File: sdk/python/kfp/dsl/_pipeline_param.py [hunk 1]
```diff
@@ -42,7 +53,7 @@ def match_serialized_pipelineparam(payload: str):
   for match in matches:
       pattern = '{{pipelineparam:op=%s;name=%s}}' % (match[0], match[1])
       param_tuples.append(PipelineParamTuple(
-                          name=sanitize_k8s_name(match[1]), 
+                          name=sanitize_k8s_name(match[1], True),
                           op=sanitize_k8s_name(match[0]),
```
Comment (Ark-kun) [inline sdk/python/kfp/dsl/_pipeline_param.py]: I'm not sure this is a proper place for sanitization. The names should probably be converted in the compiler: https://github.com/kubeflow/pipelines/blob/735e627a03629ba72a11ca89f608c1979e0c02c0/sdk/python/kfp/compiler/_op_to_template.py#L43

Comment (Ark-kun) [inline sdk/python/kfp/dsl/_pipeline_param.py]: Are you sure the op name cannot have underscores and capital letters?

Comment (numerology) [reply sdk/python/kfp/dsl/_pipeline_param.py]: It supports capital letters, but not underscores. The error I hit: `invalid spec: spec.templates[2].name: 'dataproc_create_cluster' is invalid: name must consist of alpha-numeric characters or '-', and must start with an alpha-numeric character (e.g. My-name1-2, 123-NAME)`

Comment (numerology) [reply sdk/python/kfp/dsl/_pipeline_param.py]: Agreed. I was curious why there are two sanitize functions as well.

Comment (numerology) [reply sdk/python/kfp/dsl/_pipeline_param.py]: What I intended to do here was, only bypass sanitizing `_` and capital letters when users have the need. I don't see that users would access the op names in their code in a Pythonic way so I think it's still okay to sanitize the op names.

WDYT?

Comment (Ark-kun) [reply sdk/python/kfp/dsl/_pipeline_param.py]: I just want those limitations to be addressed closer to the source of those limitations - to the compiler.
I also though that maybe the sanitization function can be simpler with only one regex.
Ideally we should move `match_serialized_pipelineparam` to the compiler.
It's OK to do that in another PR.

Comment (numerology) [reply sdk/python/kfp/dsl/_pipeline_param.py]: Thanks @Ark-kun . Will do in a following PR.

**Comment by numerology** [comment]
It seems like @k8s-ci-robot forgot this one.

**Pull request merged**